presto-main/src/main/java/com/facebook/presto/matching/MatchingEngine.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.matching;
```

Comment:
```
Move pattern matching to separate package and make it independent from
optimizer.
Thanks to that it will be possible to use pattern matching not only for
optimizer, but other components as well.
```

Even if that's the case after this PR, we don't have any guarantee it'll stay this way after it gets merged - unless we put this into a separate module. Please put this into a separate module, as this will:

1. *prove* optimizer-independence and *secure it* in the future
2. will do the same for usability out of presto-main. At this point I think using the matcher in, say, tpch connctor's tests will not be possible because of a dependency cycle.